### DISCUSSION TITLE
You cannot include a type map into itself error

### DISCUSSION THREAD

**Discussion by NCC1701M** [LuckyPennySoftware/AutoMapper#3828] [Ideas]
### Source/destination types

There are no information about which types are involed

### Mapping configuration
Always using `CreateMap<Model, ViewModel>()` and `ReverseMap`

### Expected behavior

No error!

### Actual behavior

After upgrading from AutoMapper 10 to 11 it raises the error:
`You cannot include a type map into itself error`

What does this error mean? Where and why does it happen? All mappings worked perfectly with `AutoMapper` version 10 and `AutoMapper.Extensions.Microsoft.DependencyInjection` version 8.1.1 and now this error.

There needs to be a documentation about this error or at least in the exception should be the name of the effected class.

**Comment by lbargaoanu** [comment]
A repro would help. Make a [gist](https://gist.github.com/lbargaoanu/9c7233441c3a3413cc2b9b9ebb5964a9) that we can execute and see fail.
For the error message, a PR is welcome.

**Reply by NCC1701M** [reply]
That's the problem. There are a lot of classes derived from `AutoMapper.Profile`. I can't tell which one is the effected class. And I also can't create a gist with this repo because it contains sensitive code.

At least please explain what causes this error? What should I look for?

**Reply by lbargaoanu** [reply]
You debug it the same way you debug any exception, you break in the debugger and check the local context. Without a repro as requested, that works in 10 and fails in 11, we can't help you.

**Reply by jbogard** [reply]
When you debug, you'll be able to see the `TypePair` causing the issue. @lbargaoanu I think this is just an oversight in the exception, I didn't include the type names in the exception message. This exception is thrown from `IncludeBaseTypes` and `IncludeDerivedTypes`. That should help you see what's wrong with your configuration

**Reply by NCC1701M** [reply]
> When you debug, you'll be able to see the `TypePair` causing the issue. @lbargaoanu I think this is just an oversight in the exception, I didn't include the type names in the exception message. This exception is thrown from `IncludeBaseTypes` and `IncludeDerivedTypes`. That should help you see what's wrong with your configuration

I don't use windows for developing my app and there are a lot of powershell scripts which have to be executed. I've made some attemps but I didn't get it to run, so I can't debug AutoMapper right now. It seems to be pretty cumbersome to debug it.

After you commented I was able to isolate the error to a base profile where I'm using `IncludeAllDerived`. But this class is implemented by four other files and I have to figure out which is the one causing the error.

Please fix the exception message as soon as possible.

Thanks

**Reply by jbogard** [reply]
The symbol packages are published so you should be able to debug with anything that supports symbol packages - Rider etc. You don't need the source code to debug.

**Reply by jbogard** [reply]
> Please fix the exception message as soon as possible.

We are volunteers. We do not get paid for maintaining this project. Thank you for your patience.

**Reply by NCC1701M** [reply]
> The symbol packages are published so you should be able to debug with anything that supports symbol packages - Rider etc. You don't need the source code to debug.

I'm using VSCode, don't know if OmniSharp supports this. Also couldn't find any symbol files for AutoMapper in my bin or obj folders. Right now I can't step into for example `CreateMap`

**Reply by jbogard** [reply]
No it does not because Microsoft closed-source its debugger to force people to use Visual Studio (or VS Mac).

Your best bet is to use an IDE that has a full-featured debugger

**Reply by NCC1701M** [reply]
Sorry, can't do. We're using VSCode.

**Reply by jbogard** [reply]
Then you're out of luck 🤷‍♂️

**Reply by jbogard** [reply]
One other thing you can try to debug locally is to pull down the code and reference ONLY the .csproj in your application (removing nuget package references).

Otherwise the problems you have debugging are of entirely using VS Code instead of an actual IDE. I can't help you if you can't use the right tool for the job.

**Comment by jbogard** [comment]
You can download a full IDE as a trial. Rider or VS Mac work great. The Omnisharp extension will always be intentionally under-developed by MS.
